// @fastproxy/w2/.rollup.cache/data2/workspace/allanyu/jarvis-node/packages/whistle/src/lib/plugins/get-plugins.js
import { fs, path } from '../import-pkg';
import config from '../config';
import * as util from './util';
import { getPaths } from './module-paths';
const CUSTOM_PLUGIN_PATH = config.CUSTOM_PLUGIN_PATH;
const customPluginPaths = config.customPluginPaths || [];
const notUninstallPluginPaths = config.notUninstallPluginPaths || [];
const projectPluginPaths = config.projectPluginPaths || [];
const accountPluginsPath = config.accountPluginsPath || [];
function readDir(dir, callback) {
    fs.readdir(dir, function (err, list) {
        if (!err) {
            return callback(err, list);
        }
        fs.readdir(dir, callback);
    });
}
function statFile(filepath, callback) {
    fs.stat(filepath, function (_, stat1) {
        if (stat1) {
            return callback(stat1.isFile() && stat1);
        }
        fs.stat(filepath, function (_, stat2) {
            callback(stat2 && stat2.isFile() && stat2);
        });
    });
}
function readPluginRootList(dir, callback) {
    let roots = [];
    readDir(dir, function (_, list) {
        if (!list || !list.length) {
            return callback(roots);
        }
        const handleCallback = function () {
            const orgRoots = [];
            roots = roots.filter(function (obj) {
                if (!Array.isArray(obj)) {
                    return obj;
                }
                orgRoots.push.apply(orgRoots, obj);
            });
            callback(orgRoots.concat(roots));
        };
        let count = 0;
        list.forEach(function (name, i) {
            if (util.isWhistleModule(name)) {
                roots[i] = {
                    name: name,
                    dir: dir
                };
            }
            else if (util.isOrgModule(name)) {
                ++count;
                readDir(path.join(dir, name), function (_, list2) {
                    if (list2 && list2.length) {
                        let orgList = [];
                        list2.forEach(function (pluginName) {
                            if (util.isWhistleModule(pluginName)) {
                                orgList = orgList || [];
                                orgList.push({
                                    org: name,
                                    name: pluginName,
                                    dir: dir
                                });
                            }
                        });
                        roots[i] = orgList;
                    }
                    if (--count === 0) {
                        count = -1;
                        handleCallback();
                    }
                });
            }
        });
        if (count === 0) {
            count = -1;
            handleCallback();
        }
    });
}
function readPluginModules(dir, callback, plugins, isCustom) {
    readPluginRootList(dir, function (list) {
        let len = list.length;
        if (!len) {
            return callback(plugins);
        }
        list.forEach(function (obj) {
            const dir = obj.dir;
            const dirName = obj.org ? obj.org + '/' + obj.name : obj.name;
            const root = isCustom ? path.join(dir, dirName, 'node_modules', dirName) : path.join(dir, dirName);
            statFile(path.join(root, 'package.json'), function (stats) {
                if (stats) {
                    obj.root = root;
                    obj.mtime = stats.mtime.getTime();
                }
                if (--len === 0) {
                    list.forEach(function (obj) {
                        const name = obj.name.split('.')[1];
                        if (obj.root && !plugins[name]) {
                            plugins[name] = obj;
                        }
                    });
                    callback(plugins);
                }
            });
        });
    });
}
export default function (callback) {
    const plugins = {};
    const result = {};
    const paths = getPaths();
    const count = paths.length;
    if (!count) {
        return callback(result);
    }
    const loadPlugins = function (dir, cb) {
        const isAccount = accountPluginsPath.indexOf(dir) !== -1;
        const account = isAccount ? config.account : undefined;
        const isSys = isAccount || CUSTOM_PLUGIN_PATH === dir || customPluginPaths.indexOf(dir) !== -1;
        const isProj = projectPluginPaths.indexOf(dir) !== -1;
        const notUn = notUninstallPluginPaths.indexOf(dir) !== -1;
        readPluginModules(dir, function () {
            Object.keys(plugins).filter(function (name) {
                if (util.excludePlugin(name) || result[name + ':']) {
                    return;
                }
                const obj = plugins[name];
                result[name + ':'] = {
                    account: account,
                    isSys: isSys,
                    isProj: isProj,
                    notUn: notUn,
                    path: obj.root,
                    mtime: obj.mtime
                };
            });
            cb();
        }, plugins, isSys);
    };
    let index = 0;
    const callbackHandler = function () {
        const dir = paths[++index];
        if (dir) {
            return loadPlugins(dir, callbackHandler);
        }
        callback(result);
    };
    loadPlugins(paths[index], callbackHandler);
}
