// @fastproxy/w2/.rollup.cache/data2/workspace/allanyu/jarvis-node/packages/whistle/src/lib/plugins/get-plugins-sync.js
import createDebug from 'debug';
import { fs, fse, path } from '../import-pkg';
import * as util from '../util';
import config from '../config';
import * as pluginUtil from './util';
import { getPaths } from './module-paths';
const debug = createDebug('@w2.lib.plugins.get-plugins-sync');
const CUSTOM_PLUGIN_PATH = config.CUSTOM_PLUGIN_PATH;
const customPluginPaths = config.customPluginPaths || [];
const notUninstallPluginPaths = config.notUninstallPluginPaths || [];
const projectPluginPaths = config.projectPluginPaths || [];
const accountPluginsPath = config.accountPluginsPath || [];
function readPluginModulesSync(dir, plugins = {}) {
    const isAccount = accountPluginsPath.indexOf(dir) !== -1;
    const account = isAccount ? config.account : undefined;
    const isSys = isAccount || CUSTOM_PLUGIN_PATH === dir || customPluginPaths.indexOf(dir) !== -1;
    const isProj = projectPluginPaths.indexOf(dir) !== -1;
    const notUn = notUninstallPluginPaths.indexOf(dir) !== -1;
    try {
        const list = fs.readdirSync(dir).filter(function (name) {
            if (pluginUtil.isWhistleModule(name)) {
                return true;
            }
            if (pluginUtil.isOrgModule(name)) {
                try {
                    const _dir = path.join(dir, name);
                    const org = name;
                    fs.readdirSync(_dir).forEach(function (name) {
                        if (!plugins[name] && pluginUtil.isWhistleModule(name)) {
                            const root = isSys ? path.join(_dir, name, 'node_modules', org, name) : path.join(_dir, name);
                            if (fs.existsSync(path.join(root, 'package.json'))) {
                                plugins[name] = {
                                    root: root,
                                    account: account,
                                    isSys: isSys,
                                    notUn: notUn,
                                    isProj: isProj
                                };
                            }
                        }
                    });
                }
                catch (e) { }
            }
            return false;
        });
        list.forEach(function (name) {
            if (!plugins[name]) {
                const root = isSys ? path.join(dir, name, 'node_modules', name) : path.join(dir, name);
                if (fs.existsSync(path.join(root, 'package.json'))) {
                    plugins[name] = {
                        root: root,
                        account: account,
                        isSys: isSys,
                        notUn: notUn,
                        isProj: isProj
                    };
                }
            }
        });
    }
    catch (e) { }
    return plugins;
}
export default function () {
    const plugins = {};
    const paths = getPaths();
    // throw new Error('paths length ' + paths.length);
    paths.forEach(function (dir) {
        readPluginModulesSync(dir, plugins);
    });
    debug('getPluginsSync ~ plugins: %o', Object.keys(plugins));
    const _plugins = {};
    Object.keys(plugins).forEach(function (name) {
        const simpleName = name.split('.', 2)[1];
        if (pluginUtil.excludePlugin(simpleName)) {
            return;
        }
        let dir = plugins[name];
        const account = dir.account;
        const isSys = dir.isSys;
        const isProj = dir.isProj;
        const notUn = dir.notUn;
        dir = dir.root;
        try {
            const pkgPath = path.join(dir, 'package.json');
            const pkg = fse.readJsonSync(pkgPath);
            if (pkg && pkg.version && pluginUtil.isPluginName(pkg.name)) {
                const stats = fs.statSync(pkgPath);
                const conf = pkg.whistleConfig || '';
                const tabs = conf.inspectorTabs || conf.inspectorTab || '';
                const hintList = util.getHintList(conf);
                const plugin = {
                    account: account,
                    isSys: isSys,
                    isProj: isProj,
                    notUn: notUn,
                    moduleName: pkg.name,
                    enableAuthUI: !!conf.enableAuthUI,
                    inheritAuth: !!conf.inheritAuth,
                    updateUrl: util.getUpdateUrl(conf),
                    tunnelKey: util.getTunnelKey(conf),
                    staticDir: util.getStaticDir(conf),
                    pluginVars: util.getPluginVarsConf(conf),
                    networkMenus: util.getPluginMenu(conf.networkMenus || conf.networkMenu, simpleName),
                    rulesMenus: util.getPluginMenu(conf.rulesMenus || conf.rulesMenu, simpleName),
                    valuesMenus: util.getPluginMenu(conf.valuesMenus || conf.valuesMenu, simpleName),
                    reqTab: util.getCustomTab(tabs.req, simpleName),
                    resTab: util.getCustomTab(tabs.res, simpleName),
                    tab: util.getCustomTab(tabs, simpleName),
                    comTab: util.getCustomTab(conf.composerTab, simpleName),
                    pluginHomepage: pluginUtil.getPluginHomepage(pkg),
                    openInPlugins: conf.openInPlugins ? 1 : undefined,
                    priority: parseInt(conf.priority, 10) || parseInt(pkg.pluginPriority, 10) || 0,
                    rulesUrl: util.getCgiUrl(conf.rulesUrl),
                    valuesUrl: util.getCgiUrl(conf.valuesUrl),
                    hintUrl: hintList ? undefined : util.getCgiUrl(conf.hintUrl),
                    hideShortProtocol: !!conf.hideShortProtocol,
                    hideLongProtocol: !!conf.hideLongProtocol,
                    hintList: hintList,
                    registry: util.getRegistry(pkg),
                    path: dir,
                    mtime: stats.mtime.getTime(),
                    version: pkg.version,
                    description: pkg.description,
                    homepage: pluginUtil.getHomePageFromPackage(pkg),
                    rules: util.renderPluginRules(util.trim(util.readFileSync(path.join(dir, 'rules.txt'))), pkg, simpleName),
                    _rules: util.renderPluginRules(util.trim(util.readFileSync(path.join(dir, '_rules.txt'))) ||
                        util.trim(util.readFileSync(path.join(dir, 'reqRules.txt'))), pkg, simpleName),
                    resRules: util.renderPluginRules(util.trim(util.readFileSync(path.join(dir, 'resRules.txt'))), pkg, simpleName)
                };
                plugin[util.PLUGIN_VALUES] = pluginUtil.parseValues(util.renderPluginRules(util.readFileSync(path.join(dir, '_values.txt')), pkg, simpleName));
                plugin[util.PLUGIN_MENU_CONFIG] = util.getPluginMenuConfig(conf);
                plugin[util.PLUGIN_INSPECTOR_CONFIG] = util.getPluginInspectorConfig(conf);
                _plugins[simpleName + ':'] = plugin;
            }
        }
        catch (e) {
            console.error('getPluginsSync throw error', e);
        }
    });
    return _plugins;
}
