// @fastproxy/w2/dist/biz/webui/cgi-bin/check-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const util_1 = require("../../../lib/rules/util");
const config_1 = tslib_1.__importDefault(require("../../../lib/config"));
function compare(v1, v2) {
    if (typeof v1 !== 'string') {
        return false;
    }
    if (typeof v2 !== 'string') {
        return true;
    }
    v1 = v1.split('.');
    v2 = v2.split('.');
    const v1Major = parseInt(v1[0], 10) || 0;
    const v2Major = parseInt(v2[0], 10) || 0;
    if (v1Major < v2Major) {
        return false;
    }
    if (v1Major > v2Major) {
        return true;
    }
    return parseInt(v1[1], 10) > parseInt(v2[1], 10);
}
function default_1(req, res) {
    const version = config_1.default.version;
    const doNotShowAgainVersion = util_1.properties.get('doNotShowAgainVersion');
    const latestVersion = util_1.properties.getLatestVersion();
    res.json({
        ec: 0,
        em: 'success',
        showUpdate: !config_1.default.disableUpdateTips &&
            compare(latestVersion, version) &&
            compare(latestVersion, doNotShowAgainVersion),
        version: config_1.default.version,
        latestVersion: latestVersion
    });
}
exports.default = default_1;
