// @fastproxy/w2/dist/lib/import-pkg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pkgInfo = exports.iconv = exports.checkSNI = exports.Pac = exports.bodyParser = exports.express = exports.mime = exports.forge = exports.multer = exports.wsParser = exports.hparser = exports.parseUrl = exports.url = exports.parseQuery = exports.qs = exports.getAgentServer = exports.createAgent = exports.tunnel = exports.hagent = exports.socks = exports.http2 = exports.https = exports.HttpsAgent = exports.HttpAgent = exports.http = exports.tls = exports.net = exports.dns = exports.cp = exports.PassThrough = exports.Zip = exports.parseString = exports.AdmZip = exports.Buffer = exports.LRU = exports.fse = exports.path = exports.fs = exports.assert = exports.os = exports.cluster = exports.json5 = exports.EventEmitter = exports.extend = exports.createHmac = exports.crypto = void 0;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const path_1 = tslib_1.__importDefault(require("path"));
// common about
var crypto_1 = require("crypto");
Object.defineProperty(exports, "crypto", { enumerable: true, get: function () { return tslib_1.__importDefault(crypto_1).default; } });
Object.defineProperty(exports, "createHmac", { enumerable: true, get: function () { return crypto_1.createHmac; } });
var extend_1 = require("extend");
Object.defineProperty(exports, "extend", { enumerable: true, get: function () { return tslib_1.__importDefault(extend_1).default; } });
var events_1 = require("events");
Object.defineProperty(exports, "EventEmitter", { enumerable: true, get: function () { return events_1.EventEmitter; } });
var json5_1 = require("json5");
Object.defineProperty(exports, "json5", { enumerable: true, get: function () { return tslib_1.__importDefault(json5_1).default; } });
var cluster_1 = require("cluster");
Object.defineProperty(exports, "cluster", { enumerable: true, get: function () { return tslib_1.__importDefault(cluster_1).default; } });
var os_1 = require("os");
Object.defineProperty(exports, "os", { enumerable: true, get: function () { return tslib_1.__importDefault(os_1).default; } });
var assert_1 = require("assert");
Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return tslib_1.__importDefault(assert_1).default; } });
// fs,file,process
var fs_2 = require("fs");
Object.defineProperty(exports, "fs", { enumerable: true, get: function () { return tslib_1.__importDefault(fs_2).default; } });
var path_2 = require("path");
Object.defineProperty(exports, "path", { enumerable: true, get: function () { return tslib_1.__importDefault(path_2).default; } });
var fs_extra2_1 = require("fs-extra2");
Object.defineProperty(exports, "fse", { enumerable: true, get: function () { return tslib_1.__importDefault(fs_extra2_1).default; } });
var lru_cache_1 = require("lru-cache");
Object.defineProperty(exports, "LRU", { enumerable: true, get: function () { return tslib_1.__importDefault(lru_cache_1).default; } });
var safe_buffer_1 = require("safe-buffer");
Object.defineProperty(exports, "Buffer", { enumerable: true, get: function () { return safe_buffer_1.Buffer; } });
var adm_zip_1 = require("adm-zip");
Object.defineProperty(exports, "AdmZip", { enumerable: true, get: function () { return tslib_1.__importDefault(adm_zip_1).default; } });
var xml2js_1 = require("xml2js");
Object.defineProperty(exports, "parseString", { enumerable: true, get: function () { return xml2js_1.parseString; } });
var node_native_zip2_1 = require("node-native-zip2");
Object.defineProperty(exports, "Zip", { enumerable: true, get: function () { return tslib_1.__importDefault(node_native_zip2_1).default; } });
var stream_1 = require("stream");
Object.defineProperty(exports, "PassThrough", { enumerable: true, get: function () { return stream_1.PassThrough; } });
// process about
var child_process_1 = require("child_process");
Object.defineProperty(exports, "cp", { enumerable: true, get: function () { return tslib_1.__importDefault(child_process_1).default; } });
// network about
var dns_1 = require("dns");
Object.defineProperty(exports, "dns", { enumerable: true, get: function () { return tslib_1.__importDefault(dns_1).default; } });
var net_1 = require("net");
Object.defineProperty(exports, "net", { enumerable: true, get: function () { return tslib_1.__importDefault(net_1).default; } });
var tls_1 = require("tls");
Object.defineProperty(exports, "tls", { enumerable: true, get: function () { return tslib_1.__importDefault(tls_1).default; } });
var http_1 = require("http");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return tslib_1.__importDefault(http_1).default; } });
var http_2 = require("http");
Object.defineProperty(exports, "HttpAgent", { enumerable: true, get: function () { return http_2.Agent; } });
var https_1 = require("https");
Object.defineProperty(exports, "HttpsAgent", { enumerable: true, get: function () { return https_1.Agent; } });
var https_2 = require("https");
Object.defineProperty(exports, "https", { enumerable: true, get: function () { return tslib_1.__importDefault(https_2).default; } });
var http2_1 = require("http2");
Object.defineProperty(exports, "http2", { enumerable: true, get: function () { return tslib_1.__importDefault(http2_1).default; } });
var sockx_1 = require("sockx");
Object.defineProperty(exports, "socks", { enumerable: true, get: function () { return tslib_1.__importDefault(sockx_1).default; } });
var hagent_1 = require("hagent");
Object.defineProperty(exports, "hagent", { enumerable: true, get: function () { return tslib_1.__importDefault(hagent_1).default; } });
var hagent_2 = require("hagent");
Object.defineProperty(exports, "tunnel", { enumerable: true, get: function () { return hagent_2.agent; } });
Object.defineProperty(exports, "createAgent", { enumerable: true, get: function () { return hagent_2.create; } });
Object.defineProperty(exports, "getAgentServer", { enumerable: true, get: function () { return hagent_2.getServer; } });
var querystring_1 = require("querystring");
Object.defineProperty(exports, "qs", { enumerable: true, get: function () { return tslib_1.__importDefault(querystring_1).default; } });
var querystring_2 = require("querystring");
Object.defineProperty(exports, "parseQuery", { enumerable: true, get: function () { return querystring_2.parse; } });
var url_1 = require("url");
Object.defineProperty(exports, "url", { enumerable: true, get: function () { return tslib_1.__importDefault(url_1).default; } });
var url_2 = require("url");
Object.defineProperty(exports, "parseUrl", { enumerable: true, get: function () { return url_2.parse; } });
var hparser_1 = require("hparser");
Object.defineProperty(exports, "hparser", { enumerable: true, get: function () { return tslib_1.__importDefault(hparser_1).default; } });
var ws_parser_1 = require("ws-parser");
Object.defineProperty(exports, "wsParser", { enumerable: true, get: function () { return tslib_1.__importDefault(ws_parser_1).default; } });
var multer2_1 = require("multer2");
Object.defineProperty(exports, "multer", { enumerable: true, get: function () { return tslib_1.__importDefault(multer2_1).default; } });
var node_forge_1 = require("node-forge");
Object.defineProperty(exports, "forge", { enumerable: true, get: function () { return tslib_1.__importDefault(node_forge_1).default; } });
var mime_1 = require("mime");
Object.defineProperty(exports, "mime", { enumerable: true, get: function () { return tslib_1.__importDefault(mime_1).default; } });
// framework about
var express_1 = require("express");
Object.defineProperty(exports, "express", { enumerable: true, get: function () { return tslib_1.__importDefault(express_1).default; } });
var body_parser_1 = require("body-parser");
Object.defineProperty(exports, "bodyParser", { enumerable: true, get: function () { return tslib_1.__importDefault(body_parser_1).default; } });
var node_pac_1 = require("node-pac");
Object.defineProperty(exports, "Pac", { enumerable: true, get: function () { return tslib_1.__importDefault(node_pac_1).default; } });
var sni_1 = require("sni");
Object.defineProperty(exports, "checkSNI", { enumerable: true, get: function () { return tslib_1.__importDefault(sni_1).default; } });
var iconv_lite_1 = require("iconv-lite");
Object.defineProperty(exports, "iconv", { enumerable: true, get: function () { return tslib_1.__importDefault(iconv_lite_1).default; } });
exports.pkgInfo = JSON.parse(fs_1.default.readFileSync(path_1.default.resolve(__dirname, '../../package.json'), {
    encoding: 'utf-8'
}));
