// @fastproxy/w2/dist/lib/rules/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getClientCert = exports.initRules = exports.initHeaderRules = exports.resolveResRulesFile = exports.resolveRulesFile = exports.execRulesScript = exports.getProxy = exports.disableDnsCache = exports.clearAppend = exports.resolveLocalRule = exports.lookupHost = exports.resolveBodyFilter = exports.resolveResRules = exports.resolveRules = exports.resolveRule = exports.resolvePipe = exports.resolveDisable = exports.hasReqScript = exports.resolveEnable = exports.resolveProxy = exports.resolveInternalHost = exports.resolveHost = exports.resolveSNICallback = exports.append = exports.parse = exports.Rules = exports.setGetPluginFn = exports.getPlugin = void 0;
const tslib_1 = require("tslib");
const import_pkg_1 = require("../import-pkg");
const util = tslib_1.__importStar(require("../util"));
const logger_1 = tslib_1.__importDefault(require("../util/logger"));
const fileMgr = tslib_1.__importStar(require("../util/file-mgr"));
const config_1 = tslib_1.__importDefault(require("../config"));
const util_1 = require("./util");
const rules_1 = tslib_1.__importDefault(require("./rules"));
exports.Rules = rules_1.default;
const dns_1 = tslib_1.__importDefault(require("./dns"));
const tplCache = new import_pkg_1.LRU({ max: 36 });
const rulesMgrCache = new import_pkg_1.LRU({ max: 16 });
const clientCerts = new import_pkg_1.LRU({ max: 60, maxAge: 1000 * 60 * 60 });
const rules = new rules_1.default();
const tempRules = new rules_1.default();
const cachedPacs = {};
const RULES_HEADER = 'x-whistle-rule-value';
const KV_HEADER = 'x-whistle-key-value';
const KEY_HEADER = 'x-whistle-rule-key';
const HOST_HEADER = 'x-whistle-rule-host';
const LOCALHOST = '127.0.0.1';
const resolveReqRules = rules.resolveReqRules.bind(rules);
const AUTH_RE = /^([^\\/]+)@/;
/**
 * 设置getPlugin函数
 */
const setGetPluginFn = (arg) => (exports.getPlugin = arg);
exports.setGetPluginFn = setGetPluginFn;
let parse;
exports.parse = parse;
if (config_1.default.networkMode && !config_1.default.shadowRulesMode) {
    exports.parse = parse = util.noop;
}
else {
    exports.parse = parse = (text, root, _values) => {
        if (config_1.default.pluginsMode || config_1.default.shadowRulesMode) {
            text = config_1.default.shadowRules || '';
        }
        if (config_1.default.shadowValues) {
            _values = _values
                ? (0, import_pkg_1.extend)({}, config_1.default.shadowValues, _values)
                : config_1.default.shadowValues;
        }
        rules.parse(text, root, _values);
    };
}
exports.append = rules.append.bind(rules);
exports.resolveSNICallback = rules.resolveSNICallback.bind(rules);
exports.resolveHost = rules.resolveHost.bind(rules);
exports.resolveInternalHost = rules.resolveInternalHost.bind(rules);
exports.resolveProxy = rules.resolveProxy.bind(rules);
exports.resolveEnable = rules.resolveEnable.bind(rules);
exports.hasReqScript = rules.hasReqScript.bind(rules);
exports.resolveDisable = rules.resolveDisable.bind(rules);
exports.resolvePipe = rules.resolvePipe.bind(rules);
exports.resolveRule = rules.resolveRule.bind(rules);
exports.resolveRules = resolveReqRules;
exports.resolveResRules = rules.resolveResRules.bind(rules);
exports.resolveBodyFilter = rules.resolveBodyFilter.bind(rules);
exports.lookupHost = rules.lookupHost.bind(rules);
exports.resolveLocalRule = rules.resolveLocalRule.bind(rules);
exports.clearAppend = rules.clearAppend.bind(rules);
const disableDnsCache = () => {
    rules_1.default.disableDnsCache();
};
exports.disableDnsCache = disableDnsCache;
const dnsResolve = (host, callback) => {
    return (0, dns_1.default)(host, callback || util.noop, true);
};
const PROXY_HOSTS_RE = /[?&]proxyHosts?(?:&|$)/i;
const P_HOST_RE = /[?&]host=([\w.:-]+)(?:&|$)/i;
function isProxyHost(req, proxy, host) {
    if (proxy) {
        req._proxyTunnel =
            isLineProp(proxy, 'proxyTunnel') ||
                isLineProp(host, 'proxyTunnel') ||
                util.isEnable(req, 'proxyTunnel') ||
                isProxyEnable(req, 'proxyTunnel');
        if (isLineProp(proxy, 'proxyHost')) {
            return true;
        }
    }
    return isLineProp(host, 'proxyHost');
}
function isProxyEnable(req, name) {
    let props = req._proxyProps;
    if (props === undefined) {
        props = rules.resolveProxyProps(req);
        req._proxyProps = props || null;
    }
    return props && util.isEnable(props, name);
}
function isLineProp(rule, name) {
    return rule && rule.lineProps[name];
}
const getProxy = (url, req, callback) => {
    if (!req) {
        return callback();
    }
    const reqRules = req.rules;
    req.curUrl = url;
    if (!reqRules) {
        return rules.lookupHost(req, callback);
    }
    delete reqRules.proxy;
    delete reqRules.pac;
    const pRules = req.pluginRules;
    const fRules = req.rulesFileMgr;
    if (fRules) {
        fRules._values = req._scriptValues;
    }
    const hRules = req.headerRulesMgr;
    const filter = (0, import_pkg_1.extend)(rules.resolveFilter(req), pRules && pRules.resolveFilter(req), fRules && fRules.resolveFilter(req), hRules && hRules.resolveFilter(req));
    let ignoreProxy;
    let proxy;
    let pacRule;
    const host = rules.getHost(req, pRules, fRules, hRules);
    const hostValue = util.getMatcherValue(host) || '';
    if (config_1.default.multiEnv || req._headerRulesFirst) {
        proxy =
            (pRules && pRules.resolveProxy(req, hostValue)) ||
                (hRules && hRules.resolveProxy(req, hostValue)) ||
                rules.resolveProxy(req, hostValue) ||
                (fRules && fRules.resolveProxy(req, hostValue));
    }
    else {
        proxy =
            (pRules && pRules.resolveProxy(req, hostValue)) ||
                rules.resolveProxy(req, hostValue) ||
                (fRules && fRules.resolveProxy(req, hostValue)) ||
                (hRules && hRules.resolveProxy(req, hostValue));
    }
    let proxyHostOnly = isLineProp(proxy, 'proxyHostOnly');
    let proxyHost = proxyHostOnly ||
        util.isEnable(req, 'proxyHost') ||
        isProxyEnable(req, 'proxyHost');
    if (proxy) {
        const protocol = proxy.matcher.substring(0, proxy.matcher.indexOf(':'));
        ignoreProxy =
            !filter['ignore:' + protocol] &&
                (util.isIgnored(filter, 'proxy') || util.isIgnored(filter, protocol));
        if (ignoreProxy) {
            proxy = null;
        }
        else {
            proxyHost = proxyHost || PROXY_HOSTS_RE.test(proxy.matcher);
        }
    }
    if (!ignoreProxy) {
        proxyHost = isProxyHost(req, proxy, host) || proxyHost; // 不能调换顺序
    }
    const setHost = () => {
        if (!host || req._isProxyReq) {
            return false;
        }
        reqRules.host = host;
        const hostname = util.removeProtocol(host.matcher, true);
        if (!import_pkg_1.net.isIP(hostname)) {
            req.curUrl = hostname || url;
            return rules.lookupHost(req, (err, ip) => {
                callback(err, ip, host.port, host);
            });
        }
        callback(null, hostname, host.port, host);
        return true;
    };
    const resolvePacRule = () => {
        if (pacRule != null) {
            return;
        }
        if (util.isIgnored(filter, 'pac')) {
            pacRule = false;
            return;
        }
        if (config_1.default.multiEnv || req._headerRulesFirst) {
            pacRule =
                (pRules && pRules.resolvePac(req)) ||
                    (hRules && hRules.resolvePac(req)) ||
                    rules.resolvePac(req) ||
                    (fRules && fRules.resolvePac(req)) ||
                    false;
        }
        else {
            pacRule =
                (pRules && pRules.resolvePac(req)) ||
                    rules.resolvePac(req) ||
                    (fRules && fRules.resolvePac(req)) ||
                    (hRules && hRules.resolvePac(req)) ||
                    false;
        }
        if (pacRule) {
            proxyHostOnly = isLineProp(pacRule, 'proxyHostOnly');
            proxyHost =
                proxyHost || proxyHostOnly || isLineProp(pacRule, 'proxyHost');
        }
    };
    if (host) {
        if (!proxyHost && !ignoreProxy) {
            resolvePacRule();
        }
        if (proxyHost) {
            req._phost = (0, import_pkg_1.parseUrl)(util.setProtocol(host.matcher + (host.port ? ':' + host.port : '')));
        }
        else if (!isLineProp(proxy, 'proxyFirst') &&
            !isLineProp(host, 'proxyFirst') &&
            !util.isEnable(req, 'proxyFirst') &&
            !isProxyEnable(req, 'proxyFirst')) {
            return setHost();
        }
        proxyHost = true;
        reqRules.host = host;
        req._enableProxyHost = true;
    }
    if (ignoreProxy || (proxyHostOnly && !req._phost)) {
        req.curUrl = url;
        return setHost() || rules.lookupHost(req, callback);
    }
    if (proxy) {
        if (!req._phost && P_HOST_RE.test(proxy.matcher)) {
            req._phost = (0, import_pkg_1.parseUrl)(util.setProtocol(RegExp.$1));
        }
        reqRules.proxy = proxy;
        return callback();
    }
    resolvePacRule();
    let pacUrl = util.getMatcherValue(pacRule);
    if (!pacUrl) {
        return setHost() || callback();
    }
    if (AUTH_RE.test(pacUrl)) {
        const auth = RegExp.$1;
        req._pacAuth = auth;
        pacUrl = pacUrl.substring(auth.length + 1);
    }
    pacUrl = /^https?\:\/\//.test(pacUrl)
        ? pacUrl
        : util.join(pacRule.root, pacUrl);
    let pac = cachedPacs[pacUrl];
    if (pac) {
        delete cachedPacs[pacUrl];
        cachedPacs[pacUrl] = pac;
    }
    else {
        const list = Object.keys(cachedPacs);
        if (list.length >= 10) {
            delete cachedPacs[list[0]];
        }
        cachedPacs[pacUrl] = pac = new import_pkg_1.Pac(pacUrl, dnsResolve);
    }
    reqRules.pac = pacRule;
    return pac.findWhistleProxyForURL(url.replace('tunnel:', 'https:'), (err, rule) => {
        if (rule) {
            tempRules.parse(pacRule.rawPattern + ' ' + rule);
            req.curUrl = url;
            if ((proxy = tempRules.resolveProxy(req, hostValue))) {
                proxyHost = isProxyHost(req, proxy) || proxyHost; // 不能调换顺序
                req._proxyTunnel =
                    req._proxyTunnel || isLineProp(pacRule, 'proxyTunnel');
                const protocol = proxy.matcher.substring(0, proxy.matcher.indexOf(':'));
                if (!util.isIgnored(filter, protocol)) {
                    reqRules.proxy = proxy;
                    reqRules.proxy.raw = pacRule.raw;
                }
            }
        }
        if (reqRules.proxy) {
            (!proxyHost && setHost()) || callback();
        }
        else {
            req.curUrl = url;
            setHost() || rules.lookupHost(req, callback);
        }
        logger_1.default.error(err);
    });
};
exports.getProxy = getProxy;
function tpl(str, data) {
    if (typeof str !== 'string' ||
        str.indexOf('<%') === -1 ||
        str.indexOf('%>') === -1) {
        return str + '';
    }
    const key = str;
    let fn = tplCache.get(key);
    if (!fn) {
        str = str
            .replace(/[\u2028\u2029]/g, '')
            .replace(/\t/g, ' ')
            .replace(/\r?\n|\r/g, '\t')
            .split('<%')
            .join('\u2028')
            .replace(/((^|%>)[^\u2028]*)'/g, '$1\r')
            .replace(/\u2028=(.*?)%>/g, "',$1,'")
            .split('\u2028')
            .join("');")
            .split('%>')
            .join("p.push('")
            .split('\r')
            .join("\\'");
        try {
            fn = new Function('obj', 'const p=[],print=function(){p.push.apply(p,arguments);};' +
                "with(obj){p.push('" +
                str +
                "');}return p.join('');");
        }
        catch (e) {
            fn = e;
            throw e;
        }
        finally {
            tplCache.set(key, fn);
        }
    }
    else if (typeof fn !== 'function') {
        throw fn;
    }
    return fn(data || {}).replace(/\t/g, '\n');
}
function getScriptContext(req, res, body, pattern) {
    const ip = req.clientIp || LOCALHOST;
    let ctx = req.scriptContenxt;
    if (!ctx) {
        const headers = (0, import_pkg_1.extend)(true, {}, req.headers);
        ctx = req.scriptContenxt = {
            pattern: pattern,
            version: config_1.default.version,
            port: config_1.default.port,
            uiHost: 'local.wproxy.org',
            uiPort: config_1.default.uiport,
            url: req.fullUrl,
            method: util.toUpperCase(req.method) || 'GET',
            httpVersion: req.httpVersion || '1.1',
            isLocalAddress: (_ip) => {
                return util.isLocalAddress(_ip || ip);
            },
            ip: ip,
            clientIp: ip,
            clientPort: req.clientPort,
            headers: headers,
            reqHeaders: headers,
            body: body || '',
            reqScriptData: {},
            res: null
        };
    }
    ctx.rules = [];
    ctx.values = {};
    ctx.value = req.globalValue;
    ctx.getValue = (key, onlyValues) => {
        const value = !onlyValues && req._inlineValues && req._inlineValues[key];
        return typeof value === 'string' ? value : util_1.values.get(key);
    };
    ctx.parseUrl = import_pkg_1.parseUrl;
    ctx.parseQuery = import_pkg_1.parseQuery;
    ctx.tpl = ctx.render = tpl;
    if (res) {
        ctx.statusCode = res.statusCode;
        ctx.serverIp = req.hostIp || LOCALHOST;
        ctx.resHeaders = (0, import_pkg_1.extend)(true, {}, res.headers);
    }
    else {
        ctx.statusCode = '';
        ctx.serverIp = '';
        ctx.resHeaders = '';
    }
    return ctx;
}
function getReqPayload(req, res, cb) {
    if (res) {
        return cb();
    }
    if (req.getPayload && util.hasRequestBody(req)) {
        if (typeof req._reqBody === 'string') {
            cb(req._reqBody);
        }
        else {
            req.getPayload((_, payload) => {
                cb(fileMgr.decode(payload));
            });
        }
    }
    else {
        cb();
    }
}
function execRulesScript(script, req, res, body, pattern) {
    const context = getScriptContext(req, res, body, pattern);
    if (util.execScriptSync(script, context) && Array.isArray(context.rules)) {
        return {
            rules: context.rules.join('\n').trim(),
            values: context.values
        };
    }
    return undefined;
}
exports.execRulesScript = execRulesScript;
function handleDynamicRules(script, req, res, cb) {
    util.getRuleValue(script, (list) => {
        let scriptItem;
        let index;
        let text;
        if (list) {
            index = script.scriptIndex;
            scriptItem = script.list[index];
            text = scriptItem && list[index];
        }
        if (!scriptItem || (text && util.isRulesContent(text))) {
            return cb(list && list.join('\n'));
        }
        getReqPayload(req, res, (body) => {
            const result = execRulesScript(text, req, res, body, script.rawPattern);
            list[index] = result?.rules;
            cb(list.join('\n'), result?.values);
        });
    }, null, null, null, req);
}
function resolveRulesFile(req, callback) {
    if (req.rules.G) {
        const varMap = {};
        let globalValue;
        const pList = [];
        req.rules.G.list.forEach((item) => {
            if (item.matcher[0] !== 'P') {
                if (!globalValue) {
                    globalValue = item;
                }
                return;
            }
            let value = util.getMatcherValue(item);
            const index = value.indexOf('=');
            if (index !== -1) {
                const name = value.substring(0, index);
                const plugin = (0, exports.getPlugin)?.(name);
                if (!plugin || !plugin.pluginVars) {
                    return;
                }
                value = value.substring(index + 1);
                if (value) {
                    pList.push(item);
                    const list = varMap[name] || [];
                    varMap[name] = list;
                    if (list.indexOf(value) === -1) {
                        list.push(value);
                    }
                }
            }
        });
        req._pluginVars = varMap;
        req.rules.P = pList;
        req.rules.G = globalValue;
        req.globalValue = util.getMatcherValue(globalValue);
    }
    handleDynamicRules(req.rules.rulesFile, req, null, (text, vals) => {
        if (text) {
            let rulesFileMgr = rulesMgrCache.get(text);
            if (!rulesFileMgr) {
                rulesFileMgr = new rules_1.default(vals);
                rulesFileMgr.parse(text);
                rulesMgrCache.set(text, rulesFileMgr);
            }
            rulesFileMgr._values = vals;
            req._scriptValues = vals;
            req.rulesFileMgr = rulesFileMgr;
            req.curUrl = req.fullUrl;
            text = req.rulesFileMgr.resolveReqRules(req);
        }
        // 不能放到if里面
        util.mergeRules(req, text);
        callback();
    });
}
exports.resolveRulesFile = resolveRulesFile;
const resolveResRulesFile = (req, res, callback) => {
    handleDynamicRules(req.rules && req.rules.resScript, req, res, (text, vals) => {
        text = text && text.trim();
        callback(text && {
            text: text,
            values: vals
        });
    });
};
exports.resolveResRulesFile = resolveResRulesFile;
function getValue(req, key, keep) {
    let value = req.headers[key];
    if (value) {
        if (!keep && config_1.default.strict) {
            value = null;
        }
        else {
            req.rulesHeaders[key] = value;
        }
        delete req.headers[key];
    }
    try {
        return value && decodeURIComponent(value);
    }
    catch (e) { }
    return value;
}
function initHeaderRules(req, needBodyFilters) {
    if (req._bodyFilters !== undefined) {
        return;
    }
    req._bodyFilters = null;
    req.rulesHeaders = {};
    const isPluginReq = req.isPluginReq && !req._isProxyReq;
    req._headerRulesFirst = isPluginReq;
    const rulesHeader = getValue(req, RULES_HEADER, isPluginReq);
    const hostHeader = util.trimStr(getValue(req, HOST_HEADER));
    let keyHeader = util.trimStr(getValue(req, KEY_HEADER));
    const kvHeader = getValue(req, KV_HEADER, isPluginReq);
    let ruleValue = util.trimStr(rulesHeader);
    if (hostHeader) {
        ruleValue = ruleValue + '\n' + hostHeader;
    }
    if (keyHeader) {
        keyHeader = util.trimStr(util_1.values.get(keyHeader));
        if (keyHeader) {
            ruleValue = keyHeader + '\n' + ruleValue;
        }
    }
    const curVars = rules._globalPluginVars;
    let value;
    req._globalPluginVars = {};
    if (ruleValue) {
        const rulesMgr = new rules_1.default(util.parseJSON(kvHeader));
        rulesMgr.parse(ruleValue);
        req.headerRulesMgr = rulesMgr;
        const bodyFilters = needBodyFilters && rulesMgr._rules._bodyFilters;
        if (bodyFilters && bodyFilters.length) {
            req._bodyFilters = rules._rules._bodyFilters.concat(bodyFilters);
        }
        const headerVars = rulesMgr._globalPluginVars;
        const vars = (0, import_pkg_1.extend)({}, curVars, headerVars);
        Object.keys(vars).forEach(key => {
            const plugin = (0, exports.getPlugin)?.(key);
            if (!plugin || !plugin.pluginVars) {
                return;
            }
            const headerVal = headerVars[key];
            const curVal = curVars[key];
            if (curVal && headerVal) {
                vars[key] = headerVal.concat(curVal);
            }
            value = vars[key];
            req._globalPluginVars[key] = value;
        });
    }
    else {
        Object.keys(curVars).forEach(key => {
            if (!exports.getPlugin || (0, exports.getPlugin)(key)) {
                value = curVars[key];
                req._globalPluginVars[key] = value;
            }
        });
    }
}
exports.initHeaderRules = initHeaderRules;
function initRules(req) {
    const fullUrl = req.fullUrl || util.getFullUrl(req);
    req.curUrl = fullUrl;
    initHeaderRules(req);
    if (req.headerRulesMgr) {
        if (config_1.default.multiEnv || req._headerRulesFirst) {
            req.rules = resolveReqRules(req);
            util.mergeRules(req, req.headerRulesMgr.resolveReqRules(req));
        }
        else {
            req.rules = req.headerRulesMgr.resolveReqRules(req);
            util.mergeRules(req, resolveReqRules(req));
        }
    }
    else {
        req.rules = resolveReqRules(req);
    }
    return req.rules;
}
exports.initRules = initRules;
const HTTPS_RE = /^(?:ws|http)s:\/\//;
function checkCache(cacheKey, callback) {
    let result = clientCerts.peek(cacheKey);
    if (result) {
        if (result.pending) {
            result.push(callback);
        }
        else {
            callback(result);
        }
        return true;
    }
    else {
        result = [callback];
        result.pending = true;
        clientCerts.set(cacheKey, result);
        return false;
    }
}
const getClientCert = (req, cb) => {
    if (!req) {
        return cb();
    }
    req.curUrl = req.realUrl || req.fullUrl;
    if (!HTTPS_RE.test(req.curUrl)) {
        return cb();
    }
    const pRules = req.pluginRules;
    const fRules = req.rulesFileMgr;
    const hRules = req.headerRulesMgr;
    let rule;
    if (config_1.default.multiEnv || req._headerRulesFirst) {
        rule =
            (pRules && pRules.resolveClientCert(req)) ||
                (hRules && hRules.resolveClientCert(req)) ||
                rules.resolveClientCert(req) ||
                (fRules && fRules.resolveClientCert(req));
    }
    else {
        rule =
            (pRules && pRules.resolveClientCert(req)) ||
                rules.resolveClientCert(req) ||
                (fRules && fRules.resolveClientCert(req)) ||
                (hRules && hRules.resolveClientCert(req));
    }
    if (!rule) {
        return cb();
    }
    req.rules.clientCert = rule;
    const matcherStr = rule.matcher?.substring(17);
    const matcher = matcherStr && util.parseJSON(matcherStr);
    if (!matcher) {
        return cb();
    }
    const base = util.getString(matcher.base);
    let key = util.getString(matcher.key);
    let cert = util.getString(matcher.cert);
    let cacheKey;
    try {
        if (key && cert) {
            if (base) {
                key = import_pkg_1.path.join(base, key);
                cert = import_pkg_1.path.join(base, cert);
            }
            cacheKey = 'cert\n' + key + '\n' + cert;
            if (checkCache(cacheKey, (data) => {
                if (data) {
                    cb(data[0], data[1], false, cacheKey);
                }
                else {
                    cb();
                }
            })) {
                return;
            }
            return fileMgr.readFileList([key, cert], (data) => {
                const list = clientCerts.peek(cacheKey);
                if (data[0] && data[1] && data[0].length && data[1].length) {
                    clientCerts.set(cacheKey, data);
                }
                else {
                    clientCerts.del(cacheKey);
                    data = '';
                }
                list.forEach((fn) => {
                    fn(data);
                });
            });
        }
        const pwd = util.getString(matcher.pwd || matcher.passphrase);
        let pfx = util.getString(matcher.pfx);
        if (pfx) {
            if (base) {
                pfx = import_pkg_1.path.join(base, key);
            }
            cacheKey = 'pfx\n' + pwd + '\n' + pfx;
            if (checkCache(cacheKey, (buf) => {
                if (buf) {
                    cb(pwd, buf, true, cacheKey);
                }
                else {
                    cb();
                }
            })) {
                return;
            }
            return fileMgr.readFile(pfx, (buf) => {
                const list = clientCerts.peek(cacheKey);
                if (buf && buf.length) {
                    clientCerts.set(cacheKey, buf);
                }
                else {
                    clientCerts.del(cacheKey);
                    buf = '';
                }
                list.forEach((fn) => {
                    fn(buf);
                });
            });
        }
    }
    catch (e) { }
    cb();
};
exports.getClientCert = getClientCert;
