// @fastproxy/w2/dist/lib/rules/protocols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contains = exports.isFileProxy = exports.isWebsocketProtocol = exports.isWebProtocol = exports.isResRule = exports.resetRules = exports.isBinProtocol = exports.getRules = exports.multiMatchs = exports.getRuleProto = exports.reqProtocols = exports.aliasProtocols = exports.resProtocols = exports.pureResProtocols = exports.protocols = void 0;
exports.protocols = [
    'G',
    'style',
    'host',
    'rule',
    'pipe',
    'weinre',
    'proxy',
    'https2http-proxy',
    'http2https-proxy',
    'internal-proxy',
    'pac',
    'filter',
    'ignore',
    'enable',
    'disable',
    'delete',
    'log',
    'plugin',
    'referer',
    'auth',
    'ua',
    'urlParams',
    'params',
    'resMerge',
    'replaceStatus',
    'method',
    'cache',
    'attachment',
    'forwardedFor',
    'responseFor',
    'rulesFile',
    'resScript',
    'reqDelay',
    'resDelay',
    'headerReplace',
    'reqSpeed',
    'resSpeed',
    'reqType',
    'resType',
    'reqCharset',
    'resCharset',
    'reqCookies',
    'resCookies',
    'reqCors',
    'resCors',
    'reqHeaders',
    'resHeaders',
    'trailers',
    'reqPrepend',
    'resPrepend',
    'reqBody',
    'resBody',
    'reqAppend',
    'resAppend',
    'urlReplace',
    'reqReplace',
    'resReplace',
    'reqWrite',
    'resWrite',
    'reqWriteRaw',
    'resWriteRaw',
    'cssAppend',
    'htmlAppend',
    'jsAppend',
    'cssBody',
    'htmlBody',
    'jsBody',
    'cssPrepend',
    'htmlPrepend',
    'jsPrepend',
    'cipher',
    'sniCallback'
];
const RULE_RE = /^(?:|x|xs)(?:file|rawfile|dust|tpl|jsonp):/;
const PROXY_RE = /^x?(?:socks|proxy|https?-proxy|internal-proxy|internal-https?-proxy|https2http-proxy|http2https-proxy)$/;
exports.pureResProtocols = [
    'replaceStatus',
    'cache',
    'attachment',
    'resMerge',
    'resDelay',
    'resSpeed',
    'resType',
    'resType',
    'resCharset',
    'resCookies',
    'resCors',
    'resHeaders',
    'trailers',
    'resPrepend',
    'resBody',
    'resAppend',
    'resReplace',
    'resWrite',
    'resWriteRaw',
    'cssAppend',
    'htmlAppend',
    'jsAppend',
    'cssBody',
    'htmlBody',
    'jsBody',
    'cssPrepend',
    'htmlPrepend',
    'jsPrepend',
    'responseFor'
];
exports.resProtocols = [
    'filter',
    'enable',
    'disable',
    'ignore',
    'style',
    'delete',
    'headerReplace'
].concat(exports.pureResProtocols);
const binProtocols = [
    'reqPrepend',
    'resPrepend',
    'reqBody',
    'resBody',
    'reqAppend',
    'resAppend'
];
const jsProtocols = ['jsAppend', 'jsBody', 'jsPrepend'];
const cssProtocols = ['cssAppend', 'cssBody', 'cssPrepend'];
exports.aliasProtocols = {
    ruleFile: 'rulesFile',
    ruleScript: 'rulesFile',
    rulesScript: 'rulesFile',
    reqScript: 'rulesFile',
    reqRules: 'rulesFile',
    resRules: 'resScript',
    pathReplace: 'urlReplace',
    download: 'attachment',
    skip: 'ignore',
    'http-proxy': 'proxy',
    'xhttp-proxy': 'xproxy',
    status: 'statusCode',
    hosts: 'host',
    xhost: 'host',
    html: 'htmlAppend',
    js: 'jsAppend',
    reqMerge: 'params',
    css: 'cssAppend',
    excludeFilter: 'filter',
    includeFilter: 'filter',
    P: 'G'
};
exports.reqProtocols = exports.protocols.filter(function (name) {
    return exports.pureResProtocols.indexOf(name) === -1;
});
const RULE_PROTO_RE = /^([\w.-]+):\/\//;
const getRuleProto = function (rule) {
    if (!RULE_PROTO_RE.test(rule.matcher)) {
        return;
    }
    const proto = RegExp.$1;
    const ruleProto = exports.aliasProtocols[proto];
    if (!ruleProto || ruleProto === 'filter') {
        return proto;
    }
    if (ruleProto === 'rulesFile') {
        return 'reqScript';
    }
    if (ruleProto === 'urlReplace') {
        return 'pathReplace';
    }
    if (ruleProto === 'params') {
        return 'reqMerge';
    }
    return ruleProto;
};
exports.getRuleProto = getRuleProto;
exports.multiMatchs = [
    'G',
    'ignore',
    'enable',
    'filter',
    'disable',
    'plugin',
    'delete',
    'style',
    'trailers',
    'urlParams',
    'params',
    'headerReplace',
    'reqHeaders',
    'resHeaders',
    'reqCors',
    'resCors',
    'reqCookies',
    'resCookies',
    'reqReplace',
    'urlReplace',
    'resReplace',
    'resMerge',
    'reqBody',
    'reqPrepend',
    'resPrepend',
    'reqAppend',
    'resAppend',
    'resBody',
    'htmlAppend',
    'jsAppend',
    'cssAppend',
    'htmlBody',
    'jsBody',
    'cssBody',
    'htmlPrepend',
    'jsPrepend',
    'cssPrepend',
    'rulesFile',
    'resScript'
];
function getRules() {
    return resetRules({});
}
exports.getRules = getRules;
function isBinProtocol(protocol) {
    if (binProtocols.indexOf(protocol) != -1) {
        return 1;
    }
    if (jsProtocols.indexOf(protocol) != -1) {
        return 2;
    }
    if (cssProtocols.indexOf(protocol) != -1) {
        return 3;
    }
    return undefined;
}
exports.isBinProtocol = isBinProtocol;
function resetRules(rules) {
    exports.protocols.forEach(function (protocol) {
        rules[protocol] = [];
    });
    rules._localRule = [];
    rules._bodyFilters = [];
    rules._clientCerts = [];
    rules.rule.isRuleProto = true;
    return rules;
}
exports.resetRules = resetRules;
function isResRule(protocol) {
    return exports.resProtocols.indexOf(protocol) != -1;
}
exports.isResRule = isResRule;
function isWebProtocol(protocol) {
    return protocol == 'http:' || protocol == 'https:';
}
exports.isWebProtocol = isWebProtocol;
function isWebsocketProtocol(protocol) {
    return protocol == 'ws:' || protocol == 'wss:';
}
exports.isWebsocketProtocol = isWebsocketProtocol;
function isFileProxy(protocol) {
    return RULE_RE.test(protocol);
}
exports.isFileProxy = isFileProxy;
function contains(name) {
    if (exports.protocols.indexOf(name) != -1 ||
        exports.aliasProtocols[name] ||
        PROXY_RE.test(name)) {
        return true;
    }
    name += ':';
    return (isWebsocketProtocol(name) ||
        isWebProtocol(name) ||
        isFileProxy(name) ||
        name == 'tunnel:');
}
exports.contains = contains;
