// @fastproxy/w2/.rollup.cache/data2/workspace/allanyu/jarvis-node/packages/whistle/src/lib/rules/protocols.js
export const protocols = [
    'G',
    'style',
    'host',
    'rule',
    'pipe',
    'weinre',
    'proxy',
    'https2http-proxy',
    'http2https-proxy',
    'internal-proxy',
    'pac',
    'filter',
    'ignore',
    'enable',
    'disable',
    'delete',
    'log',
    'plugin',
    'referer',
    'auth',
    'ua',
    'urlParams',
    'params',
    'resMerge',
    'replaceStatus',
    'method',
    'cache',
    'attachment',
    'forwardedFor',
    'responseFor',
    'rulesFile',
    'resScript',
    'reqDelay',
    'resDelay',
    'headerReplace',
    'reqSpeed',
    'resSpeed',
    'reqType',
    'resType',
    'reqCharset',
    'resCharset',
    'reqCookies',
    'resCookies',
    'reqCors',
    'resCors',
    'reqHeaders',
    'resHeaders',
    'trailers',
    'reqPrepend',
    'resPrepend',
    'reqBody',
    'resBody',
    'reqAppend',
    'resAppend',
    'urlReplace',
    'reqReplace',
    'resReplace',
    'reqWrite',
    'resWrite',
    'reqWriteRaw',
    'resWriteRaw',
    'cssAppend',
    'htmlAppend',
    'jsAppend',
    'cssBody',
    'htmlBody',
    'jsBody',
    'cssPrepend',
    'htmlPrepend',
    'jsPrepend',
    'cipher',
    'sniCallback'
];
const RULE_RE = /^(?:|x|xs)(?:file|rawfile|dust|tpl|jsonp):/;
const PROXY_RE = /^x?(?:socks|proxy|https?-proxy|internal-proxy|internal-https?-proxy|https2http-proxy|http2https-proxy)$/;
export const pureResProtocols = [
    'replaceStatus',
    'cache',
    'attachment',
    'resMerge',
    'resDelay',
    'resSpeed',
    'resType',
    'resType',
    'resCharset',
    'resCookies',
    'resCors',
    'resHeaders',
    'trailers',
    'resPrepend',
    'resBody',
    'resAppend',
    'resReplace',
    'resWrite',
    'resWriteRaw',
    'cssAppend',
    'htmlAppend',
    'jsAppend',
    'cssBody',
    'htmlBody',
    'jsBody',
    'cssPrepend',
    'htmlPrepend',
    'jsPrepend',
    'responseFor'
];
export const resProtocols = [
    'filter',
    'enable',
    'disable',
    'ignore',
    'style',
    'delete',
    'headerReplace'
].concat(pureResProtocols);
const binProtocols = [
    'reqPrepend',
    'resPrepend',
    'reqBody',
    'resBody',
    'reqAppend',
    'resAppend'
];
const jsProtocols = ['jsAppend', 'jsBody', 'jsPrepend'];
const cssProtocols = ['cssAppend', 'cssBody', 'cssPrepend'];
export const aliasProtocols = {
    ruleFile: 'rulesFile',
    ruleScript: 'rulesFile',
    rulesScript: 'rulesFile',
    reqScript: 'rulesFile',
    reqRules: 'rulesFile',
    resRules: 'resScript',
    pathReplace: 'urlReplace',
    download: 'attachment',
    skip: 'ignore',
    'http-proxy': 'proxy',
    'xhttp-proxy': 'xproxy',
    status: 'statusCode',
    hosts: 'host',
    xhost: 'host',
    html: 'htmlAppend',
    js: 'jsAppend',
    reqMerge: 'params',
    css: 'cssAppend',
    excludeFilter: 'filter',
    includeFilter: 'filter',
    P: 'G'
};
export const reqProtocols = protocols.filter(function (name) {
    return pureResProtocols.indexOf(name) === -1;
});
const RULE_PROTO_RE = /^([\w.-]+):\/\//;
export const getRuleProto = function (rule) {
    if (!RULE_PROTO_RE.test(rule.matcher)) {
        return;
    }
    const proto = RegExp.$1;
    const ruleProto = aliasProtocols[proto];
    if (!ruleProto || ruleProto === 'filter') {
        return proto;
    }
    if (ruleProto === 'rulesFile') {
        return 'reqScript';
    }
    if (ruleProto === 'urlReplace') {
        return 'pathReplace';
    }
    if (ruleProto === 'params') {
        return 'reqMerge';
    }
    return ruleProto;
};
export const multiMatchs = [
    'G',
    'ignore',
    'enable',
    'filter',
    'disable',
    'plugin',
    'delete',
    'style',
    'trailers',
    'urlParams',
    'params',
    'headerReplace',
    'reqHeaders',
    'resHeaders',
    'reqCors',
    'resCors',
    'reqCookies',
    'resCookies',
    'reqReplace',
    'urlReplace',
    'resReplace',
    'resMerge',
    'reqBody',
    'reqPrepend',
    'resPrepend',
    'reqAppend',
    'resAppend',
    'resBody',
    'htmlAppend',
    'jsAppend',
    'cssAppend',
    'htmlBody',
    'jsBody',
    'cssBody',
    'htmlPrepend',
    'jsPrepend',
    'cssPrepend',
    'rulesFile',
    'resScript'
];
export function getRules() {
    return resetRules({});
}
export function isBinProtocol(protocol) {
    if (binProtocols.indexOf(protocol) != -1) {
        return 1;
    }
    if (jsProtocols.indexOf(protocol) != -1) {
        return 2;
    }
    if (cssProtocols.indexOf(protocol) != -1) {
        return 3;
    }
    return undefined;
}
export function resetRules(rules) {
    protocols.forEach(function (protocol) {
        rules[protocol] = [];
    });
    rules._localRule = [];
    rules._bodyFilters = [];
    rules._clientCerts = [];
    rules.rule.isRuleProto = true;
    return rules;
}
export function isResRule(protocol) {
    return resProtocols.indexOf(protocol) != -1;
}
export function isWebProtocol(protocol) {
    return protocol == 'http:' || protocol == 'https:';
}
export function isWebsocketProtocol(protocol) {
    return protocol == 'ws:' || protocol == 'wss:';
}
export function isFileProxy(protocol) {
    return RULE_RE.test(protocol);
}
export function contains(name) {
    if (protocols.indexOf(name) != -1 ||
        aliasProtocols[name] ||
        PROXY_RE.test(name)) {
        return true;
    }
    name += ':';
    return (isWebsocketProtocol(name) ||
        isWebProtocol(name) ||
        isFileProxy(name) ||
        name == 'tunnel:');
}
