// @fastproxy/w2/dist/lib/plugins/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.excludePlugin = exports.getPluginHomepage = exports.parseValues = exports.getHomePageFromPackage = exports.isWhistleModule = exports.isPluginName = exports.isOrgModule = void 0;
const tslib_1 = require("tslib");
const util = tslib_1.__importStar(require("../util"));
const config_1 = tslib_1.__importDefault(require("../config"));
const protocols = tslib_1.__importStar(require("../rules/protocols"));
const ORG_RE = /^@[\w-]+$/;
const WHISLTE_PLUGIN_RE = /^whistle\.[a-z\d_\-]+$/;
const HTTP_RE = /^https?:\/\//i;
const PLUGIN_NAME_RE = /^(?:@[\w-]+\/)?whistle\.[a-z\d_\-]+$/;
function isOrgModule(name) {
    return ORG_RE.test(name);
}
exports.isOrgModule = isOrgModule;
const isPluginName = function (name) {
    return PLUGIN_NAME_RE.test(name);
};
exports.isPluginName = isPluginName;
function isWhistleModule(name) {
    return WHISLTE_PLUGIN_RE.test(name);
}
exports.isWhistleModule = isWhistleModule;
function getHomePageFromPackage(pkg) {
    if (HTTP_RE.test(pkg.homepage)) {
        return pkg.homepage;
    }
    return extractUrl(pkg.repository) || '';
}
exports.getHomePageFromPackage = getHomePageFromPackage;
function extractUrl(repository) {
    if (!repository || repository.type != 'git' || typeof repository.url !== 'string') {
        return;
    }
    let url = repository.url.replace(/^git\+/i, '');
    if (!HTTP_RE.test(url)) {
        url = url.replace(/^git@([^:]+):/, 'http://$1/');
    }
    return url.replace(/\.git\s*$/i, '');
}
const parseValues = function (val) {
    if (val) {
        val = util.parseJSON(val);
    }
    if (!val) {
        return '';
    }
    Object.keys(val).forEach(function (key) {
        val[key] = util.toString(val[key]);
    });
    return val;
};
exports.parseValues = parseValues;
const getPluginHomepage = function (pkg) {
    const url = pkg.pluginHomepage || pkg.pluginHomePage;
    return typeof url === 'string' ? url : undefined;
};
exports.getPluginHomepage = getPluginHomepage;
const excludePlugin = function (name) {
    if (protocols.contains(name) || (config_1.default.allowPluginList.length && config_1.default.allowPluginList.indexOf(name) === -1)) {
        return true;
    }
    return config_1.default.blockPluginList.length && config_1.default.blockPluginList.indexOf(name) !== -1;
};
exports.excludePlugin = excludePlugin;
