// @fastproxy/w2/dist/lib/util/parse-url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const URL_RE = /^([a-z0-9.+-]+:)\/\/([^/?#]+)(\/[^?#]*)?(\?[^#]*)?(#.*)?$/i;
const HOST_RE = /^(.+)(?::(\d*))$/;
const BRACKET_RE = /^\[|\]$/g;
function default_1(url) {
    if (!URL_RE.test(url)) {
        return (0, url_1.parse)(url);
    }
    const protocol = RegExp.$1;
    const host = RegExp.$2;
    const pathname = RegExp.$3 || '/';
    const search = RegExp.$4;
    const hash = RegExp.$5 || null;
    let port = null;
    let hostname = host;
    if (HOST_RE.test(host)) {
        hostname = RegExp.$1;
        port = RegExp.$2;
    }
    return {
        protocol: protocol,
        slashes: true,
        auth: null,
        host: host,
        port: port,
        hostname: hostname.replace(BRACKET_RE, ''),
        hash: hash,
        search: search || null,
        query: search ? search.substring(1) : null,
        pathname: pathname,
        path: pathname + search,
        href: url
    };
}
exports.default = default_1;
