// @fastproxy/w2/dist/lib/util/perf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setProxy = exports.procData = void 0;
const now = Date.now();
const preData = {
    now: now,
    totalHttpRequests: 0,
    totalWsRequests: 0,
    totalTunnelRequests: 0,
    totalAllHttpRequests: 0,
    totalAllWsRequests: 0
};
const memUsage = process.memoryUsage();
let maxCpuElap = 0;
exports.procData = {
    memUsage: memUsage,
    uptime: 0,
    cpuPercent: '0.0%',
    startupTime: now,
    updateTime: now,
    httpRequests: 0,
    allHttpRequests: 0,
    wsRequests: 0,
    allWsRequests: 0,
    tunnelRequests: 0,
    totalHttpRequests: 0,
    totalWsRequests: 0,
    totalTunnelRequests: 0,
    totalAllHttpRequests: 0,
    totalAllWsRequests: 0,
    httpQps: 0,
    tunnelQps: 0,
    wsQps: 0,
    totalQps: 0,
    maxQps: 0,
    maxAllQps: 0,
    maxRss: memUsage.rss,
    maxCpu: '0',
    maxQpsTime: now,
    maxAllQpsTime: now,
    maxRssTime: now,
    maxCpuTime: now,
    allHttpQps: 0,
    allWsQps: 0,
    totalAllQps: 0
};
let startTime = typeof process.hrtime === 'function' && process.hrtime();
let startUsage = typeof process.cpuUsage === 'function' && process.cpuUsage();
let proxy;
function secNSec2ms(secNSec) {
    if (Array.isArray(secNSec)) {
        return secNSec[0] * 1000 + secNSec[1] / 1000000;
    }
    return secNSec / 1000;
}
if (startTime !== false && startUsage !== false) {
    setInterval(function () {
        const elapTime = process.hrtime(startTime);
        const elapUsage = process.cpuUsage(startUsage);
        startTime = process.hrtime();
        startUsage = process.cpuUsage();
        const elapTimeMS = secNSec2ms(elapTime) || 1;
        const elapUserMS = secNSec2ms(elapUsage.user);
        const elapSystMS = secNSec2ms(elapUsage.system);
        const cpuElap = (100 * (elapUserMS + elapSystMS)) / elapTimeMS;
        const curTime = Date.now();
        exports.procData.cpuPercent = cpuElap.toFixed(1) + '%';
        exports.procData.memUsage = process.memoryUsage();
        exports.procData.updateTime = curTime;
        if (cpuElap > maxCpuElap) {
            maxCpuElap = cpuElap;
            exports.procData.maxCpu = cpuElap.toFixed(1) + '%';
            exports.procData.maxCpuTime = curTime;
        }
        if (exports.procData.memUsage.rss > exports.procData.maxRss) {
            exports.procData.maxRss = exports.procData.memUsage.rss;
            process.maxRssTime = curTime;
        }
        proxy && proxy.emit('perfDataChange', exports.procData);
    }, 3000);
    setInterval(function () {
        const curTime = Date.now();
        const costTime = curTime - preData.now || 1;
        const newHttpReqs = exports.procData.totalHttpRequests - preData.totalHttpRequests;
        const newTunnelReqs = exports.procData.totalTunnelRequests - preData.totalTunnelRequests;
        const newWsReqs = exports.procData.totalWsRequests - preData.totalWsRequests;
        const newHttpUIReqs = exports.procData.totalAllHttpRequests - preData.totalAllHttpRequests;
        const newWsUIReqs = exports.procData.totalAllWsRequests - preData.totalAllWsRequests;
        preData.now = curTime;
        preData.totalHttpRequests = exports.procData.totalHttpRequests;
        preData.totalTunnelRequests = exports.procData.totalTunnelRequests;
        preData.totalWsRequests = exports.procData.totalWsRequests;
        preData.totalAllHttpRequests = exports.procData.totalAllHttpRequests;
        preData.totalAllWsRequests = exports.procData.totalAllWsRequests;
        exports.procData.uptime = curTime - now;
        exports.procData.httpQps = Math.floor((newHttpReqs * 100000) / costTime);
        exports.procData.tunnelQps = Math.floor((newTunnelReqs * 100000) / costTime);
        exports.procData.wsQps = Math.floor((newWsReqs * 100000) / costTime);
        exports.procData.allHttpQps = Math.floor((newHttpUIReqs * 100000) / costTime);
        exports.procData.allWsQps = Math.floor((newWsUIReqs * 100000) / costTime);
        const totalQps = exports.procData.httpQps + exports.procData.tunnelQps + exports.procData.wsQps;
        const totalAllQps = exports.procData.allHttpQps + exports.procData.allWsQps + exports.procData.tunnelQps;
        exports.procData.totalQps = totalQps;
        exports.procData.totalAllQps = totalAllQps;
        if (exports.procData.maxQps < totalQps) {
            exports.procData.maxQps = totalQps;
            exports.procData.maxQpsTime = curTime;
        }
        if (exports.procData.maxAllQps < totalAllQps) {
            exports.procData.maxAllQps = totalAllQps;
            exports.procData.maxAllQpsTime = curTime;
        }
    }, 1000);
}
const setProxy = function (p) {
    proxy = p;
};
exports.setProxy = setProxy;
